### PR TITLE
I420toRGB and I420toBGR operations for CPU plugin

### PR DESCRIPTION
**Pull request by vladislav-volkov**
### Details:
 - I420toRGB and I420toBGR operations for CPU plugin

### Tickets:
 - CVS-71214

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp [hunk 1]
```diff
@@ -0,0 +1,1113 @@
...
+
+}   // namespace
+
+MKLDNNColorConvertNode::Converter::Converter(MKLDNNNode *node, const ColorFormat & colorFormat)
+    : _node(node)
+    , _colorFormat(colorFormat) {
+}
+
+InferenceEngine::Precision MKLDNNColorConvertNode::Converter::inputPrecision(size_t idx) const {
+    return _node->getParentEdgeAt(idx)->getMemory().GetDescWithType<BlockedMemoryDesc>()->getPrecision();
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp:942]: better to use `get*EdgesAtPort`
it's necessary for child edges when node has several output port
unlikely `ColorConvert` will have more that one output ports but better avoid it in advance

also it's not necessary to call `GetDescWithType` to get precision enough `getDesc()`

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp [hunk 2]
```diff
@@ -0,0 +1,1113 @@
...
+    std::string errorMessage;
+    std::tie(algorithm, errorMessage) = getAlgorithmFor(op);
+    if (algorithm == Algorithm::Default)
+        IE_THROW(NotImplemented) << errorMessage;
+}
+
+void MKLDNNColorConvertNode::getSupportedDescriptors() {}
+
+void MKLDNNColorConvertNode::initSupportedPrimitiveDescriptors() {
+    if (supportedPrimitiveDescriptors.empty()) {
```
Comment (mandrono) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp:980]: maybe? like in other nodes
```suggestion
   if (!supportedPrimitiveDescriptors.empty())
        return;
```

**Comment by mandrono** [comment]
Is it possible to add tests for these nodes?

**Comment by vladislav-volkov** [comment]
> Is it possible to add tests for these nodes?

We already have tests. Use filetrs *_i420_* and *I420* in cpuFuncTests.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp (hunk 2 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp]: fixed

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    static bool isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op, std::string& errorMessage) noexcept;
+
+private:
+    void initSupportedNV12Impls();
+    void initSupportedI420Impls();
+
+private:
+    using ConverterBuilder = std::function<Converter*(MKLDNNNode *)>;
+    using SupportedImpls = multidim_map<impl_desc_type, Algorithm, InferenceEngine::Precision::ePrecision, bool, ConverterBuilder>;
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.h]: Comment added

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    std::unique_ptr<Converter> _impl;
+    SupportedImpls _supportedImpls;
+};
+
+class MKLDNNColorConvertNode::Converter {
+public:
+    using PrimitiveDescs = std::vector<std::tuple<std::vector<PortConfigurator>,
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.h]: Comments added

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.h [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    using ColorFormat = std::array<uint8_t, 3>;
+
+    Converter(MKLDNNNode *node, const ColorFormat & colorFormat);
+    virtual ~Converter() = default;
+    InferenceEngine::Precision inputPrecision(size_t idx) const;
+    InferenceEngine::Precision outputPrecision(size_t idx) const;
+    const void * input(size_t idx) const;
+    void * output(size_t idx) const;
+    const VectorDims & inputDims(size_t idx) const;
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.h]: "inputPrecision" and "input" is shorter than "getInputPrecision" and "getInputDataPtr".
I think, here "get" prefix is redundant.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp [hunk 6]
```diff
@@ -0,0 +1,1113 @@
...
+using namespace InferenceEngine;
+using namespace mkldnn::impl::utils;
+using namespace mkldnn::impl::cpu::x64;
+using namespace Xbyak;
+
+namespace MKLDNNPlugin {
+
+namespace {
+
+std::tuple<Algorithm, std::string> getAlgorithmFor(const std::shared_ptr<const ngraph::Node>& op) {
```
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp]: Without map of lambdas current implementation is shorter and readable.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp (hunk 1 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp]: getDesc() is private. So I can't use it.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp]: Type name added to error message.

**Comment by dmitry-gorokhov** [comment]
As we discussed here we need to fix leftovers from https://github.com/openvinotoolkit/openvino/pull/8628 as first step

**Comment by vladislav-volkov** [comment]
> As we discussed here we need to fix leftovers from #8628 as first step
All leftovers were fixed.

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp (hunk 1 above)
Comment (dmitry-gorokhov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp]: @vladislav-volkov It is definitely public: https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/src/mkldnn_memory.h#L52. So lets refactor this

File: src/plugins/intel_cpu/src/utils/jit_kernel.cpp [hunk 7]
```diff
@@ -15,6 +18,12 @@ namespace {
 template<typename RegType>
 using registers = std::array<std::reference_wrapper<const RegType>, 16>;
 
+bool isRegAllocable(int id) {
+    return id != abi_param1.getIdx()          // function argument
+            && id != Operand::Code::RSP       // stack pointer
+            && id != Operand::Code::RBP;      // frame pointer
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/utils/jit_kernel.cpp:24]: 1. I guess RBP is wifely used in other JIT kernels for computations. Why we protect it here?
2. JIT kernel might use abi_param1 to pass call arguments. How we preserve this register?

File: src/plugins/intel_cpu/src/utils/jit_kernel.hpp [hunk 8]
```diff
@@ -577,51 +837,82 @@ void jit_kernel::foreach(const B & begin,
...
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// shared_reg
+
+template<typename Reg>
+shared_reg<Reg> make_shared(Reg & reg, jit_kernel & kernel) {
+    std::shared_ptr<Reg> ptr(&reg, [&kernel](Reg *preg) {
+        try {
+            kernel.free(*preg);
+        } catch(...) {}
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/utils/jit_kernel.hpp:885]: Why we do nothing in catch section?

File: src/plugins/intel_cpu/src/utils/jit_kernel.hpp [hunk 9]
```diff
@@ -632,6 +923,22 @@ void boolean_expression<T>::cmp(const Xbyak::Label & exit) const {
             _kernel.je(exit, Xbyak::CodeGenerator::T_NEAR);
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/utils/jit_kernel.hpp:923]: BTW shouldn't it be `_kernel.jne` here? If not some comment required for sure.

File: src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp [hunk 10]
```diff
@@ -0,0 +1,347 @@
+// Copyright (C) 2021 Intel Corporation
```
Comment (dmitry-gorokhov) [inline src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp:1]: please remove `mkldnn_` from the name of the file.

File: src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+    size_t c = 3;
+    size_t d = 4;
+
+    size_t result64 = 0;
+    jit_scalar_variable_test_kernel<size_t> kernel64;
+    jit_scalar_variable_test_kernel<size_t>::Params args64 = { a, b, c, d, &result64 };
+    kernel64(args64);
+
+    uint32_t result32 = 0;
+    jit_scalar_variable_test_kernel<uint32_t> kernel32;
```
Comment (dmitry-gorokhov) [inline src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp:85]: Is it possible to cover signed and floating point precisions as well or it is not supported by the `jit_kernel` yet?

**Comment by dmitry-gorokhov** [comment]
@mandrono We still need your approval here

**Review by mandrono — APPROVED** [review]

File: src/plugins/intel_cpu/src/utils/jit_kernel.hpp (hunk 9 above)
Comment (vladislav-volkov) [reply src/plugins/intel_cpu/src/utils/jit_kernel.hpp:923]: This jump is used to go to the exit point.
This is used to generate the following code:
> cmp(a, b);
je(else);
/// here a != b
L(else);
/// here a == b

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp (hunk 1 above)
Comment (vladislav-volkov) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_color_convert_node.cpp]: done

File: src/plugins/intel_cpu/src/utils/jit_kernel.cpp (hunk 7 above)
Comment (vladislav-volkov) [reply src/plugins/intel_cpu/src/utils/jit_kernel.cpp:24]: 1. The RBP removed from this expression.
2. I think abi_param1 register should not be changed during code generation. This is useful for restoring the original input values passed to the jit kernel.

File: src/plugins/intel_cpu/src/utils/jit_kernel.hpp (hunk 8 above)
Comment (vladislav-volkov) [reply src/plugins/intel_cpu/src/utils/jit_kernel.hpp:885]: We can't throw exception form the destructor. I can only suppress exceptions. ):

File: src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp (hunk 11 above)
Comment (vladislav-volkov) [reply src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp:85]: Signed int data types are supported but the single floating point operations aren't implemented.
Added tests for expressions with signed data types.

File: src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp (hunk 10 above)
Comment (vladislav-volkov) [reply src/tests/unit/cpu/mkldnn_jit_kernel_test.cpp]: Test file was renamed.

**Review by dmitry-gorokhov — APPROVED** [review]

File: src/plugins/intel_cpu/src/utils/jit_kernel.hpp (hunk 9 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/src/utils/jit_kernel.hpp:923]: I guess we need to add comment that. Right now it looks a little bit confusing.

**Pull request merged**